### PR TITLE
Feat: Adding limit field to template and Taks, overriding the CLI args from task.

### PR DESCRIPTION
**Pull request by Gnyblast**
Fix for https://github.com/semaphoreui/semaphore/issues/1357 and making override of the CLI args from task run dialog:
- Added limits fields to the template form and runs task dialog
- Created DB fields for this new field to be stored
- Changed Models to support parsing from database or api json
- Overriding implemented for the limit from task to template to enable pre-run changes to take effect
- Added language element for "limit"
- Fixed a JS error that was getting thrown due to item elements was being tried to set while item was undefined
- Made run dialog to display the args defined on template to make it obvious for the end-user to see what's going to be ran
- Made it also override the args by the task to template, so that if you change the args on pre-run it's really a override.

### CODE REVIEW

File: services/tasks/LocalJob.go
Comment (fiftin) [inline services/tasks/LocalJob.go]: This code required because task can be triggered automatically via Integration or Schedule. Not only from web interface.

So, template's `hosts_limit` must be used on the backend too if task's `hosts_limit` is empty.

Comment (Gnyblast) [reply services/tasks/LocalJob.go]: I implemted a new function called IsCLIArgOverriden that compares the task and template cli args one by one to each other and manages to decide if an template argument is being overriden by a task cli argument. The what it does it, replacing the CLI argument of `templateExtraArgs` with task argument that is targeting to override it. This way task arguments are redundent after the comparison because we are managing to override them to templateExtraArgs variable and use that only. This method should be covering any way of running a template.

**Comment by Gnyblast** [comment]
Not sure why codacy fails though...

**Comment by fiftin** [comment]
Thank you! Reviewing

**Comment by Gnyblast** [comment]
Thanks waiting for the response, we are kind of waiting this for migration.

File: web/package.json
Comment (fiftin) [inline web/package.json]: I think this change shouldn't be in the PR ;)

File: services/tasks/LocalJob_test.go
Comment (fiftin) [inline services/tasks/LocalJob_test.go:21]: Single argument can not contain space which separates key and value. It must be 2 separate arguments. Otherwise `ansible-playbook` doesn't accept this argument (tested).

**Review by fiftin — CHANGES_REQUESTED** [review]

File: services/tasks/LocalJob_test.go
Comment (Gnyblast) [reply services/tasks/LocalJob_test.go:21]: I didn't get full what you mean? Do you mean that `--ssh-extra-args` will only work if separator is `=` sign? Like: `-ssh-extra-args="-p 3222"`? and won't work with `--ssh-extra-args "-p 3222"`? If that's the case, how do you prevent this on template CLI args declaration already? And this is not a user mistake if that's the case? On the other hand this piece of code only is to determine if they are passing same argument from task CLI args, so that it can be overriden, I don't really touch CLI args how they look like, regex is just checking either first `=` or first white space and splits it from there, so it can compare the passed option, `--ssh-extra-args` in this case.

File: web/package.json
Comment (Gnyblast) [reply web/package.json]: Yeah my bad fixing this: wasn't sure what port vue uses by default, so I created the container exposing 3001 for it and had to add it there.

Comment (Gnyblast) [reply web/package.json]: I pushed the fix for this.

File: services/tasks/LocalJob_test.go
Comment (fiftin) [reply services/tasks/LocalJob_test.go:21]: @Gnyblast yes, it is true. Argument `--ssh-extra-args "-p 3222"` doesn't work in Semaphore.

You can try to call ansible CLI with params `ansible-playbook '--ssh-extra-args "-p 3222"'` and you will got error. But following command will work: `ansible-playbook '--ssh-extra-args' '-p 3222'`

Semaphore runs ansible by command like `exec.Command("ansible-playbook", []string{"--ssh-extra-args \"-p 3222\""})` so it will not work. But following code will work: `exec.Command("ansible-playbook", []string{"--ssh-extra-args", "-p 3222"})`

Comment (Gnyblast) [reply services/tasks/LocalJob_test.go:21]: as I said before this piece of code here is just to test if the parser regex can split argument and it's value correctly by delimiter which is either `=` sign or a space. I don't do any formatting as you can see in the code below:
```
	for _, taskArg := range taskExtraArgs {
		for tai, tmplArg := range templateExtraArgs {
			ok, err := isCLIArgsOverridden(tmplArg, taskArg)
			if err != nil {
				t.Log(err.Error())
			}

			if ok {
				templateExtraArgs = append(templateExtraArgs[:tai], templateExtraArgs[tai+1:]...)
				break
			}

			if slices.Contains(templateExtraArgs, taskArg) {
				templateExtraArgs = append(templateExtraArgs[:tai], templateExtraArgs[tai+1:]...)
			}

		}
	}
```
first piece with `isCLIArgsOverridden` is the logic split them by delimiter and decide if it's same argument key with different value, the:
1- if yes, that means it's an override and delete the one on `templateExtraArgs` since `taskExtraArgs` will override that specific arguments
2- If not and if there are duplicate, then again delete the on in `templateExtraArgs` since `taskExtraArgs` already have a duplicate of it.

So if user does a mistake on how to pass a CLI argument, which they can do it now with the current code, it's a user mistake and should be fixed by passing it correctly but using `=` sign or escaping correctly. Nothing changes there.

**Comment by fiftin** [comment]
Hi @Gnyblast I reviewing.

File: db/Task.go
Comment (fiftin) [inline db/Task.go:30]: How this related to `--limit`?

Comment (fiftin) [reply db/Task.go:30]: - I can not merge this PR, I don't understand what this PR do. 
- This changes can break current API usage. 
- We have flag `Task.AllowOverrideArgsInTask` but it is ignored in the PR.
- `api-docs.yml` reformatted. What reason? This change not related to the PR.

You need `--limit` right? Lets create separate PR for `--limit` and I will merge it after review.

**Review by fiftin — CHANGES_REQUESTED** [review]

File: db/Task.go
Comment (Gnyblast) [reply db/Task.go:30]: > I can not merge this PR, I don't understand what this PR do.

Might be beause of too much gap between each review and forgetting the context.

> We have flag Task.AllowOverrideArgsInTask but it is ignored in the PR.

Can you tell me where exactly I remove that so it's ignored? As far as I see you just use that variable to display the CLI args dialog on new task dialog.

> api-docs.yml reformatted. What reason? This change not related to the PR.

Well that's probably because of my yaml formatter, that could be fixed but doesn't matter anymore.

Just note this PR to somewhere, at some point you will understand the current logic failure of CLI args override and it will be helpful.

**Pull request closed without merging**
